### PR TITLE
Vectorize reduction when reducing on fastest striding dimension

### PR DESCRIPTION
**Pull request by zasdfgbnm**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#36709 Vectorize reduction when reducing on fastest striding dimension**

Differential Revision: [D21083393](https://our.internmc.facebook.com/intern/diff/D21083393)

### CODE REVIEW

**Comment by zasdfgbnm** [comment]
Performance on P100:

```python
import torch
print(torch.__version__)
print()

for i in range(1000):
    torch.arange(10000, device='cuda')

def benchmark(dtype, i):
    numel = 2 ** i
    a = torch.zeros(numel, device='cuda', dtype=dtype)
    torch.cuda.synchronize()
    %timeit a.sum(dtype=dtype); torch.cuda.synchronize()

for dtype in [torch.int8, torch.half, torch.float, torch.double]:
    print(dtype)
    for i in range(18, 30):
        benchmark(dtype, i)
    print()
```

Before
```
1.5.0a0+3bbb36e

torch.int8
28.9 µs ± 77.9 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
31.2 µs ± 36.5 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
33.1 µs ± 476 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
36.3 µs ± 387 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
40.8 µs ± 183 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
63.2 µs ± 190 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
96.8 µs ± 114 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
195 µs ± 267 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
369 µs ± 246 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
665 µs ± 1.22 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
1.3 ms ± 1.15 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
2.61 ms ± 2.06 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

torch.float16
29.9 µs ± 60.3 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
32.3 µs ± 41.8 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
34.5 µs ± 68.3 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
38.1 µs ± 214 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
50.2 µs ± 55.4 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
68.4 µs ± 49.8 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
106 µs ± 107 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
229 µs ± 420 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
407 µs ± 239 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
717 µs ± 1.05 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
1.42 ms ± 1.26 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
2.79 ms ± 841 ns per loop (mean ± std. dev. of 7 runs, 100 loops each)

torch.float32
28.7 µs ± 118 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
30.7 µs ± 43.8 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
34.2 µs ± 68.8 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
47.4 µs ± 223 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
61.4 µs ± 69.6 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
90.7 µs ± 110 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
151 µs ± 44 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
314 µs ± 274 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
533 µs ± 126 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
964 µs ± 437 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
1.89 ms ± 307 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
3.72 ms ± 954 ns per loop (mean ± std. dev. of 7 runs, 100 loops each)

torch.float64
30 µs ± 69.1 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
34.1 µs ± 76.1 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
47 µs ± 51.9 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
61.6 µs ± 52.1 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
89.2 µs ± 62.3 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
145 µs ± 58.1 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
255 µs ± 432 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
516 µs ± 185 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
952 µs ± 150 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
1.82 ms ± 497 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
3.61 ms ± 1.32 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
7.2 ms ± 1.57 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

After
```
1.5.0a0+3bbb36e

torch.int8
27.9 µs ± 46.4 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
30.3 µs ± 72.1 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
31.7 µs ± 94.1 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
34.5 µs ± 119 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
38.3 µs ± 383 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
48.3 µs ± 123 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
63.3 µs ± 251 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
115 µs ± 116 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
177 µs ± 173 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
301 µs ± 1.46 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
559 µs ± 205 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
1.07 ms ± 758 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)

torch.float16
28.3 µs ± 52.6 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
30.3 µs ± 39.4 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
32.4 µs ± 324 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
36.2 µs ± 130 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
45.7 µs ± 64.4 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
60 µs ± 346 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
88.8 µs ± 416 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
163 µs ± 303 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
272 µs ± 611 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
489 µs ± 122 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
942 µs ± 1.36 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
1.82 ms ± 1.26 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

torch.float32
27.8 µs ± 36.9 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
29.9 µs ± 77.9 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
33.4 µs ± 159 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
46 µs ± 68.5 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
59.9 µs ± 61.3 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
87.7 µs ± 112 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
144 µs ± 77.4 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
269 µs ± 124 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
488 µs ± 522 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
921 µs ± 432 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
1.82 ms ± 401 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
3.58 ms ± 730 ns per loop (mean ± std. dev. of 7 runs, 100 loops each)

torch.float64
29 µs ± 81.3 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
33.1 µs ± 89.7 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
45.8 µs ± 76.6 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
60.4 µs ± 98.1 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
88.3 µs ± 55.7 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
144 µs ± 51.1 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
255 µs ± 93.4 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
527 µs ± 236 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
955 µs ± 98.1 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
1.82 ms ± 128 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
3.62 ms ± 2.26 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
7.23 ms ± 1.83 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

**Comment by zasdfgbnm** [comment]
failure unrelated

**Comment by ngimel** [comment]
This looks good, can you please add tests checking the handling of non-vectorized heads/tails because that code is essentially untested? I'm pretty sure all our current testcases (and your benchmarks) are running on power-of-2 aligned tensors.

File: aten/src/ATen/native/cuda/Reduce.cuh [hunk 1]
```diff
@@ -373,28 +390,86 @@ struct ReduceOp {
...
+
+    // combine accumulators
+    #pragma unroll
+    for (int i = 1; i < 4; i++) {
+      value_list[0] = ops.combine(value_list[0], value_list[i]);
+    }
+    return value_list[0];
+  }
+
+  C10_DEVICE arg_t thread_reduce(arg_t value, const scalar_t* data, index_t end) const {
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/Reduce.cuh:451]: why did thread_reduce need to change (e.g. get an `end` argument? It's always called with `end=config.num_inputs` as far as I can see.

Comment (zasdfgbnm) [reply aten/src/ATen/native/cuda/Reduce.cuh:451]: You are right, I was originally putting `vectorized_thread_reduce` inside `thread_reduce`, therefore the change. But then I moved it out, and forget to change it back.

Comment (zasdfgbnm) [reply aten/src/ATen/native/cuda/Reduce.cuh]: fixed

File: aten/src/ATen/native/cuda/Reduce.cuh [hunk 2]
```diff
@@ -373,28 +390,86 @@ struct ReduceOp {
...
+  C10_DEVICE arg_t header_reduce(arg_t value, const scalar_t* &data, index_t &end) const {
+    const int align_bytes = alignof(at::native::memory::aligned_vector<scalar_t, 4>);
+    const int shift = ((uint64_t)data) % align_bytes / sizeof(scalar_t);
+    if (shift > 0) {
+      data -= shift;
+      end += shift;
+      if(threadIdx.x >= shift && config.should_reduce_tail()){
+        value = ops.reduce(value, data[threadIdx.x], threadIdx.x - shift);
+      }
+      end -= blockDim.x;
```
[the remark this answers is not in the document]
Comment (zasdfgbnm) [reply aten/src/ATen/native/cuda/Reduce.cuh:406]: This is wrong, it is only correct when `blockDim.x` == 32 I think. I will fix it.

[the remark this answers is not in the document]
Comment (zasdfgbnm) [reply aten/src/ATen/native/cuda/Reduce.cuh]: fixed

File: aten/src/ATen/native/cuda/Reduce.cuh [hunk 3]
```diff
@@ -373,28 +390,86 @@ struct ReduceOp {
...
+      data += blockDim.x;
+    }
+    return value;
+  }
+
+  C10_DEVICE arg_t vectorized_thread_reduce(arg_t value, const scalar_t* data, index_t end) const {
+    using load_t = at::native::memory::aligned_vector<scalar_t, 4>;
```
[the remark this answers is not in the document]
Comment (zasdfgbnm) [reply aten/src/ATen/native/cuda/Reduce.cuh]: fixed

**Comment by zasdfgbnm** [comment]
@ngimel All comments resolved, and test added.

**Review by ngimel — APPROVED** [review]

**Pull request closed without merging**